lib/spack/spack/build_systems/autotools.py

```diff
@@ -110,70 +109,49 @@ def _do_patch_config_files(self):
         else:
             config_arch = 'local'
 
-        my_config_files = {'guess': None, 'sub': None}
         config_files = {'guess': None, 'sub': None}
         config_args = {'guess': [], 'sub': [config_arch]}
 
         for config_name in config_files.keys():
             config_file = 'config.{0}'.format(config_name)
-            if os.path.exists(config_file):
-                # First search the top-level source directory
-                my_config_files[config_name] = os.path.abspath(config_file)
-            else:
-                # Then search in all sub directories recursively.
-                # We would like to use AC_CONFIG_AUX_DIR, but not all packages
-                # ship with their configure.in or configure.ac.
-                config_path = next((os.path.abspath(os.path.join(r, f))
-                                    for r, ds, fs in os.walk('.') for f in fs
-                                    if f == config_file), None)
-                my_config_files[config_name] = config_path
-
-            if my_config_files[config_name] is not None:
+            my_config_file_list = find('.', config_file)
+
+            my_config_files = []
+            for conf in my_config_file_list:
                 try:
-                    config_path = my_config_files[config_name]
+                    config_path = conf
```

Comment:
Can we turn this `try/except` block into a small function used for filtering? My understanding of what this method does overall is:
1. Search for all `config.guess` and `config.sub` in the current stage
2. Check which ones don't work and need patching
3. Copy newer files in their place

I think we can get rid of most of these nested for loops and conditionals and write something along the lines above. Something like:
```python
def _runs_ok(script):
    try:
        script_args = ...
        check_call([script] + script_args, stdout=PIPE, stderr=PIPE)
    except Exception as e:
        tty.debug(e)
        return False
    return True

# Find candidates for patching
to_be_pached = fs.find(self.stage.path, files=['config.sub', 'config.guess'], recursive=True)
# Filter out what already works
to_be_patched = [f for f in to_be_patched if not _runs_ok(f)]
```
Let me know if I misunderstood something or if you think there's something wrong or not clear in what I say.

Reply:
fixed.